modules/packages/LinearAlgebra.chpl

```diff
@@ -2063,6 +2323,137 @@ proc kron(A: [?ADom] ?eltType, B: [?BDom] eltType) {
   return C;
 }
 
+/*
+  Matrix exponential using Pade approximation. This method returns N*N matrix which
+  is Matrix exponential of `A`
+
+  Args:
+  A: Expects an N*N square matrix
+  UseExactOneNorm: boolean value specifying if the onenorm has to be exact.
+  Defaults to true.
+
+  TODO: Need to write a method to find approximate norm.
+  Currently only supports for UseExactOneNorm = true.
+
+  .. note::
+  [1] Awad H. Al-Mohy and Nicholas J. Higham (2009)
+  "A New Scaling and Squaring Algorithm for the Matrix Exponential."
+  SIAM Journal on Matrix Analysis and Applications.
+  31 (3). pp. 970-989. ISSN 1095-7162
+
+  :throws LinearAlgebraError: If Input Matrix is not Square Matrix.
+*/
+proc expm(A: [], param UseExactOneNorm=true) throws {
+
+  if A.rank != 2 || A.domain.dim(0) != A.domain.dim(1) then {
+    throw new LinearAlgebraError('Square Matrix Expected');
+  }
+
+  var h = new ExpmPadeHelper(A, UseExactOneNorm);
+
+  // Try Pade order 3
+  var eta_1 = max(abs(h.comp_Loosed4()), abs(h.comp_Loosed6()));
+  if eta_1 < 1.495585217958292e-002 then {
+    var (U,V) = h.pade3();
+    return solvePQ(U, V);
+  }
+
+  // Try Pade order 5
+  var eta_2 = max(abs(h.comp_Exactd4()), abs(h.comp_Loosed6()));
+  if eta_2 < 2.539398330063230e-001 then {
+    var (U,V) = h.pade5();
+    return solvePQ(U, V);
+  }
+
+  // Try Pade order 7 and 9
+  var eta_3 = max(abs(h.comp_Exactd6()), abs(h.comp_Loosed8()));
+  if eta_3 < 9.504178996162932e-001 then {
+    var (U,V) = h.pade7();
+    return solvePQ(U, V);
+  }
+
+  if eta_3 < 2.097847961257068e+000 then {
+    var (U,V) = h.pade9();
+    return solvePQ(U, V);
+  }
+
+  // Use Pade order 13.
+  var eta_4 = max(abs(h.comp_Loosed8()), abs(h.comp_Loosed10()));
+  var eta_5 = min(eta_3, eta_4);
+  var theta_13 = 4.25;
+
+  var s = 0;
+  // choose smallest s such that s>=0 and 2**(-s) eta_5 <= theta_13
+  if eta_5 == 0 then {
+    // Nilpotent special case
+    s = 0;
+  }
+  else {
+    s = max(ceil(log2(eta_5 / theta_13)): int, 0);
+  }
+  var mat = h.pade13_scaled(s);
+  var U = mat[0];
+  var V = mat[1];
+  var X = solvePQ(U, V);
+  // According to the paper X = r_13(A)^(2^s): achieved by repeated squaring.
+  for i in 1..s {
+    X = dot(X, X);
+  }
+  return X;
+}
+
+/*
+This method finds P Q Matrices, where P = U + V and Q = -U + V
+it then returns X where Q*X = P.
+*/
+private proc solvePQ(U: [?D], V: [D]){
+  var P = U + V;
+  var Q = -U + V;
+  var res : [D] U.eltType;
+  var b : [D.dim(0)] U.eltType;
+
+  // Need to rewrite solve function
+  // to solve for A*X = B where B
+  // is also a matrix
+  for j in D.dim(1) {
+    for i in D.dim(0) {
+      b[i] = P[i,j];
+    }
+    b = solve(Q, b);
+    for i in D.dim(0) {
+      res[i,j] = b[i];
+    }
+  }
+  return res;
+}
+
+/*
```

Comment:
It's great that we got this for free! I'd also recommend adding `sincos` that returns both `sin` and `cos`, since that avoids duplicating work.